graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,327 @@
+    // (uncached) or the default if specified in the constructor.
+    // (Non-object fields by default are assumed to inherit their
+    // cacheability from their parents. But on the other hand, while
+    // root non-object fields can get explicit directives from their
+    // definition on the Query/Mutation object, if that doesn't exist
+    // then there's no parent field that would assign the default
+    // maxAge, so we do it here.)
+    //
+    // You can disable this on a non-root field by writing
+    // `@cacheControl(inheritMaxAge: true)` on it. If you do this,
+    // then its children will be treated like root paths, since there
+    // is no parent maxAge to inherit.
+
+    if (!fieldPolicy.hasMaxAge()
+        && ((unwrappedFieldType instanceof GraphQLCompositeType && !inheritMaxAge)
+        || parameters.getExecutionStepInfo().getPath().isRootPath())) {
+      fieldPolicy.restrict(defaultMaxAge);
+    }
+
+    return new SimpleInstrumentationContext<ExecutionResult>() {
+      @Override
+      public void onCompleted(ExecutionResult result, Throwable t) {
+        state.overallPolicy.restrict(fieldPolicy);
+        super.onCompleted(result, t);
+      }
+    };
+  }
+
+  enum CacheControlScope {
+    PUBLIC,
+    PRIVATE
+  }
+
+  private static class CacheControlState implements InstrumentationState {
+    public final CacheControlPolicy overallPolicy = new CacheControlPolicy();
+  }
+
+  private static class CacheControlPolicy {
+    @Nullable
+    private Integer maxAge;
+    @Nullable
+    private CacheControlScope scope = CacheControlScope.PUBLIC;
+
+    void restrict(CacheControlPolicy policy) {
+      if (policy.maxAge != null && (maxAge == null || policy.maxAge < maxAge)) {
+        this.maxAge = policy.maxAge;
+      }
+
+      if (policy.scope != null && (scope == null || !scope.equals(CacheControlScope.PRIVATE))) {
+        this.scope = policy.scope;
+      }
+    }
+
+    void restrict(CacheControlDirective directive) {
+      if (directive.maxAge != null && (maxAge == null || directive.maxAge < maxAge)) {
+        this.maxAge = directive.maxAge;
+      }
+
+      if (directive.scope != null && (scope == null || !scope.equals(CacheControlScope.PRIVATE))) {
+        this.scope = directive.scope;
+      }
+    }
+
+    void restrict(Integer maxAge) {
+      if (this.maxAge == null || maxAge < this.maxAge) {
+        this.maxAge = maxAge;
+      }
+    }
+
+    void replace(CacheControlDirective directive) {
+      if (directive.maxAge != null) {
+        this.maxAge = directive.maxAge;
+      }
+
+      if (directive.scope != null) {
+        this.scope = directive.scope;
+      }
+    }
+
+    void replace(@Nullable CacheControlScope scope) {
+      if (scope != null) {
+        this.scope = scope;
+      }
+    }
+
+    public Optional<String> maybeAsString() {
+      Integer maxAgeValue = maxAge == null ? 0 : maxAge;
+      if (maxAgeValue.equals(0)) {
+        return Optional.empty();
```

Comment:
That looks dangerous. Isn't `0` a valid `maxAge` for stuff that expires straight away?

Reply:
Here's where I got this logic: https://github.com/apollographql/apollo-server/blob/fcb000c6f9295290ee5f3a28267251690170e6df/packages/apollo-server-core/src/cachePolicy.ts#L26

Reply:
i think the assumption here is that in an HTTP context, things should not be assumed to be cacheable unless they claim to be cacheable, so we can equate "no cache policy" with "not cacheable". It's possible this assumption is wrong though!